File: INLBreadcrumbs/INLBreadcrumb.m
//
//  INLBreadcrumb.m
//  INLBreadcrumbs
//
//  Created by Tomas Hakel on 02/12/2015.
//  Copyright © 2015 Inloop, s.r.o. See LICENSE.md
//

#import "INLBreadcrumb.h"
#import "INLBreadcrumbCompatibleController.h"
#import "INLNavbarTitleView.h"
#import "INLBreadcrumbManager.h"
#import "UIView+INLViewWithClass.h"
#import "MethodSwizzler.h"

@implementation INLBreadcrumb

+(instancetype)breadcrumbWithController:(UIViewController<INLBreadcrumbCompatibleController> *)controller {
	return [[INLBreadcrumb alloc] initWithController:controller];
}

-(instancetype)initWithController:(UIViewController<INLBreadcrumbCompatibleController> *)controller {
	return [self initWithController:controller manager:[INLBreadcrumbManager defaultManager]];
}

+(instancetype)breadcrumbWithController:(UIViewController<INLBreadcrumbCompatibleController> *)controller manager:(INLBreadcrumbManager *)manager {
	return [[INLBreadcrumb alloc] initWithController:controller manager:manager ? manager : [INLBreadcrumbManager defaultManager]];
}

-(instancetype)initWithController:(UIViewController<INLBreadcrumbCompatibleController> *)controller manager:(INLBreadcrumbManager *)manager {
	if (self = [super init]) {
		self.controller = controller;
		self.manager = manager;
		self.cancelButtonTitle = manager.cancelButtonTitle ? manager.cancelButtonTitle : @"Cancel";
		self.breadcrumbIndicator = manager.breadcrumbIndicator ? manager.breadcrumbIndicator : @" ▾";
		self.titleView = [UIView viewWithClass:[INLNavbarTitleView class]];
		[self updateTitle];
		
		__weak typeof(self) weakSelf = self;
		self.titleView.onTitleTapped = ^{
			[weakSelf onTitleTapped];
		};
	}
	return self;
}

-(BOOL)isTopBreadcrumb {
	return [self.manager breadcrumbs].count == 0 || [self.manager breadcrumbs][0].controller == self.controller;
}

-(void)onTitleTapped {
	if ([self isTopBreadcrumb]) {
		return;
	}
	UIAlertController * popup = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
	
	NSArray<INLBreadcrumb *> * breadcrumbs = [self.manager breadcrumbs];
	for (NSInteger i = breadcrumbs.count-2; 0 <= i; --i) {
		[popup addAction:[UIAlertAction actionWithTitle:breadcrumbs[i].title style:UIAlertActionStyleDefault
			handler:^(UIAlertAction * action) {
				[self.controller.navigationController popToViewController:(UIViewController *)breadcrumbs[i].controller animated:YES];
			}]];		
	}
	[popup addAction:[UIAlertAction actionWithTitle:self.cancelButtonTitle style:UIAlertActionStyleCancel handler:nil]];
	
	popup.popoverPresentationController.sourceView = self.controller.navigationItem.titleView;
	popup.popoverPresentationController.sourceRect = CGRectMake(0, 0,
																self.controller.navigationItem.titleView.frame.size.width,
																self.controller.navigationItem.titleView.frame.size.height);
	[self.controller presentViewController:popup animated:YES completion:nil];
}

-(INLNavbarTitleView *)titleView {
	return (INLNavbarTitleView *)self.controller.navigationItem.titleView;
}

-(void)setTitleView:(INLNavbarTitleView *)titleView {
	self.controller.navigationItem.titleView = titleView;
}

-(void)updateTitle {
	NSString * title = self.title;
	if (![self isTopBreadcrumb] && self.breadcrumbIndicator) {
		title = [self.title stringByAppendingString:self.breadcrumbIndicator];
	}
	self.titleView.title = title;
}

-(NSString *)title {
	return self.controller.title;
}

-(void)setBreadcrumbIndicator:(NSString *)breadcrumbIndicator {
	_breadcrumbIndicator = breadcrumbIndicator;
	[self updateTitle];
}

@end

File: INLBreadcrumbs/INLBreadcrumb.swift
//
//  INLBreadcrumb.swift
//  INLBreadcrumbs
//
//  Created by Petr Tomicek on 13.12.16.
//  Copyright © 2016 Inloop, s.r.o. All rights reserved.
//

import Foundation
import UIKit

public class INLBreadcrumb
{
    unowned var controller: INLBreadcrumbViewController
    unowned var manager: INLBreadcrumbManager
    var titleView: INLNavbarTitleView
    
    private var cancelButtonTitle: String
    private var breadcrumbIndicator: String
    
    
    class func breadcrumb(with controller: INLBreadcrumbViewController) -> INLBreadcrumb {
        
        return INLBreadcrumb(with: controller)
    }
    
    convenience init(with controller: INLBreadcrumbViewController) {
        
        self.init(with: controller, manager: INLBreadcrumbManager.shared)
    }
    
    class func breadcrumb(with controller: INLBreadcrumbViewController, manager: INLBreadcrumbManager) -> INLBreadcrumb {
        
        return INLBreadcrumb(with: controller, manager: manager)
    }
    
    init(with controller: INLBreadcrumbViewController, manager: INLBreadcrumbManager) {
        
        self.controller = controller
        self.manager = manager
        
        if let title = manager.cancelButtonTitle {
            
            self.cancelButtonTitle = title
        }
        else {
            
            self.cancelButtonTitle = "Cancel"
        }
        
        if let indicator = manager.breadcrumbIndicator {
            
            self.breadcrumbIndicator = indicator
        }
        else {
            
            self.breadcrumbIndicator = " ▾"
        }
        
        let nib = Bundle(for: INLNavbarTitleView.self)
        let titleview = (UINib(nibName: "INLNavbarTitleView", bundle: nib).instantiate(withOwner: nil, options: nil)[0])
        
        self.titleView = titleview as! INLNavbarTitleView
        self.updateTitle()
        self.titleView.titleTappedDelegate = { () in
            
            self.onTitleTapped()
        }
        self.controller.navigationItem.titleView = self.titleView
        
    }
    
    func isTopBreadcrumb() -> Bool {
        
        return ((self.manager.breadcrumbs().count == 0) || (self.manager.breadcrumbs()[0].controller == self.controller))
    }
    
    func updateTitle() {
        
        var _title = self.getTitle()
        
        if !self.isTopBreadcrumb() {
            
            _title = self.getTitle()?.appending(breadcrumbIndicator)
        }
        self.titleView.title = _title
    }
    
    func getTitle() -> String? {
        
        return self.controller.title
    }
    
    private func onTitleTapped() {
        
        if self.isTopBreadcrumb()
        {
            return
        }
        
        let popup = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        
        var breadcrumbs = self.manager.breadcrumbs()
        
        for i in stride(from: breadcrumbs.count-2, to: -1, by: -1)  {
            
            popup.addAction(UIAlertAction(title: breadcrumbs[i].getTitle(), style: .default, handler: { (action) in
                
                self.controller.navigationController?.popToViewController(breadcrumbs[i].controller, animated: true)
            }))
        }
        
        popup.addAction(UIAlertAction(title: self.cancelButtonTitle, style: .cancel, handler: nil))
        
        popup.popoverPresentationController?.sourceView = self.controller.navigationItem.titleView
        
        popup.popoverPresentationController?.sourceRect = CGRect(x: 0, y: 0, width: self.controller.navigationItem.titleView?.frame.size.width ?? 0, height: self.controller.navigationItem.titleView?.frame.size.height ?? 0)
        
        //[popup setAssociatedProperty:@selector(setupBreadcrumbs) value:@YES];
        
        self.controller.present(popup, animated: true, completion: nil)
    }
    
    func setBreadcrumbIndicator(indicator: String) {
        
        self.breadcrumbIndicator = indicator
        self.updateTitle()
    }
}

